### PR TITLE
#732 - Teams

### LINKED ISSUE #732: Users - Add Teams
### Current Features

Nothing allows groupings of projects or people together.

### Desired Additional Features

Add the construct of a team with a designated team leader and team members, allow association of projects to a team.

### Screenshots (as needed)

_No response_

### PR DESCRIPTION
**Pull request by anthonybernardi**
## Changes

This is an initial PR. I've only done the schema so far and some types set up. Bare bones of two basic endpoints too.

## Notes

A lot more to do but I just want to make sure that I'm on the right track

## Test Cases

still need to manually test the backend stuff and make sure it's working

## To Do

- [ ] finish up endpoints
- [ ] make hooks and stuff on front end for the end points
- [ ] make forms on front end for the teams

Closes #732

### CODE REVIEW

File: src/backend/prisma/schema.prisma [hunk 1]
```diff
@@ -61,6 +61,17 @@ model User {
   projectLead             WBS_Element[]    @relation(name: "projectLead")
   projectManager          WBS_Element[]    @relation(name: "projectManager")
   Session                 Session[]
+  teams                   Team[]           @relation(name: "teamsAsMember")     
+  teamAsLead              Team?            @relation(name: "teamAsLead")
+}
+
+model Team {
+  teamId         Int            @id @default(autoincrement())
```
Comment (jamescd18) [inline src/backend/prisma/schema.prisma:69]: UUID?

File: src/backend/prisma/schema.prisma [hunk 2]
```diff
@@ -61,6 +61,17 @@ model User {
...
+  teamName       String
+  leaderId       Int
+  leader         User           @relation(name: "teamAsLead", fields:[leaderId], references: [userId]) 
+  projects       Project[]      
+  members        User[]         @relation(name: "teamsAsMember")
```
Comment (jamescd18) [inline src/backend/prisma/schema.prisma:74]: Should ask stakeholders about this but other things we could add:
- weekly meeting info: day of week, time, location, Zoom link
- intro / onboarding / team charter doc link: all business sub-teams have a charter doc, some have an onboarding doc, and I've heard some engineering teams have an intro / onboarding doc

File: src/frontend/app/app-authenticated/app-authenticated.tsx [hunk 3]
```diff
@@ -28,6 +29,7 @@ const AppAuthenticated: React.FC = () => {
             <Route path={routes.CHANGE_REQUESTS} component={ChangeRequests} />
             <Route path={routes.SETTINGS} component={Settings} />
             <Route path={routes.HELP} component={HelpPage} />
+            <Route path={routes.TEAMS} component={TeamsPage} />
```
Comment (jamescd18) [inline src/frontend/app/app-authenticated/app-authenticated.tsx:32]: Personally just think it should go right before settings just cuz those are single sorta static pages

File: src/frontend/pages/TeamsPage/TeamsPage.tsx [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+ * This file is part of NER's PM Dashboard and licensed under GNU AGPLv3.
+ * See the LICENSE file in the repository root folder for details.
+ */
+
+import styles from './TeamsPage.module.css';
+
+const TeamsPage: React.FC = () => {
+  return (
+    <div className={`pt-5 ${styles.page_not_found}`}>
+      <h1 className={styles.title}>Uh Oh</h1>
```
Comment (jamescd18) [inline src/frontend/pages/TeamsPage/TeamsPage.tsx:11]: Love this

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Few things here and there. You'll get into the fun stuff soon so we should talk at some point about who can see what where, and who can edit what where.

File: src/backend/prisma/schema.prisma (hunk 1 above)
Comment (anthonybernardi) [reply src/backend/prisma/schema.prisma:69]: true

File: src/frontend/app/app-authenticated/app-authenticated.tsx (hunk 3 above)
Comment (anthonybernardi) [reply src/frontend/app/app-authenticated/app-authenticated.tsx:32]: do you mean on the sidebar?

File: src/backend/functions/teams.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  ...descBullet,
+  id: descBullet.descriptionId,
+  dateDeleted: descBullet.dateDeleted ?? undefined
+});
+
+const teamsTransformer = (team: Prisma.TeamGetPayload<typeof relationArgs>): Team => {
+  if (team === null) throw new TypeError('Team not found');
+
+  return {
+    teamId: team.teamId ?? undefined,
```
Comment (jamescd18) [inline src/backend/functions/teams.ts:51]: Why `?? undefined`? Isn't this a field that's always required? Plus didn't we already check team being null, so at this point this field should be guaranteed to be there, right?

File: src/backend/functions/teams.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+// Fetch all teams
+const getAllTeams: ApiRouteFunction = async () => {
+  const teams = await prisma.team.findMany(relationArgs);
+  return buildSuccessResponse(teams.map(teamsTransformer));
+};
+
+// make a team with the given leader and name
+const createTeam: ApiRouteFunction = async (params: { leaderId: string; teamName: string }) => {
+  const { teamName } = params;
```
Comment (jamescd18) [inline src/backend/functions/teams.ts:71]: Does this work correctly for post body payload? I wanna say this looks like path parameters or query string parameters, but of which aren't the ideal implementation here

File: src/utils/src/types/team-types.ts [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+export interface Team {
+  teamId: number;
+  teamName: string;
+  leader: User;
+  members: User[];
+  projects: ProjectPreview[];
+}
+
+export interface TeamPreview {
```
Comment (jamescd18) [inline src/utils/src/types/team-types.ts:17]: Not sure how lovely / shitty it is, but I think we might maybe actually be able to get away with using TypeScript Utility types: Pick<Team, "teamId", "teamName"> or a different one / implementation if I'm forgetting

**Review by jamescd18 — CHANGES_REQUESTED** [review]
Loving it so far!

File: src/utils/src/types/team-types.ts (hunk 7 above)
Comment (anthonybernardi) [reply src/utils/src/types/team-types.ts:17]: ooh cool

File: src/backend/functions/teams.ts (hunk 6 above)
Comment (anthonybernardi) [reply src/backend/functions/teams.ts:71]: true, gonna delete this for now just to get the basic stuff out

File: src/backend/functions/teams.ts (hunk 5 above)
Comment (anthonybernardi) [reply src/backend/functions/teams.ts:51]: true

File: src/frontend/pages/TeamsPage/TeamsPage.tsx (hunk 4 above)
Comment (anthonybernardi) [reply src/frontend/pages/TeamsPage/TeamsPage.tsx:11]: yahoo

File: src/frontend/layouts/sidebar/sidebar.tsx [hunk 8]
```diff
@@ -31,6 +32,11 @@ const Sidebar: React.FC = () => {
       icon: faExchangeAlt,
       route: routes.CHANGE_REQUESTS
     },
+    {
+      name: 'Teams',
+      icon: faUsers,
```
[the remark this answers is not in the document]
Comment (anthonybernardi) [reply src/frontend/layouts/sidebar/sidebar.tsx:37]: will do

**Comment by anthonybernardi** [comment]
Teams page and icon ui:
image

File: src/backend/prisma/schema.prisma (hunk 2 above)
Comment (anthonybernardi) [reply src/backend/prisma/schema.prisma:74]: true true

**Comment by anthonybernardi** [comment]
So my plan for the description field is that the users write it in markdown (and maybe there's a preview button and a link to a simple markdown guide), then to store it in the database we have to replace every line break with a unique character/string (i.e., ###THIS IS A LINE BREAK###) because otherwise postgres replaces them with just a normal space. Then when the front end retrieves it to render it again you just replace all instances of the unique character with \n and then use a package to display it as markdown

**Comment by jamescd18** [comment]
Are we not able to use \n as the special character? Does Postgres scrub that out?

**Comment by anthonybernardi** [comment]
> Are we not able to use \n as the special character? Does Postgres scrub that out?

yeah that would probably work, I just tested it with a longer one

**Pull request merged**